File: C/dijkstra/deepseek-r1-distill-llama-70b.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define INF INT_MAX

// Structure to represent a graph edge
typedef struct Edge {
    int dest;
    int weight;
    struct Edge* next;
} Edge;

// Structure to represent a priority queue element
typedef struct PriorityQueue {
    int distance;
    int node;
    struct PriorityQueue* next;
} PriorityQueue;

// Structure to represent a graph
typedef struct Graph {
    int numNodes;
    Edge** adjLists;
    PriorityQueue* priorityQueue;
    int* distances;
    int* visited;
} Graph;

// Function to create a new edge
Edge* createEdge(int dest, int weight) {
    Edge* newEdge = (Edge*) malloc(sizeof(Edge));
    newEdge->dest = dest;
    newEdge->weight = weight;
    newEdge->next = NULL;
    return newEdge;
}

// Function to create a new priority queue element
PriorityQueue* createPriorityQueueElement(int distance, int node) {
    PriorityQueue* newElement = (PriorityQueue*) malloc(sizeof(PriorityQueue));
    newElement->distance = distance;
    newElement->node = node;
    newElement->next = NULL;
    return newElement;
}

// Function to insert an element into the priority queue
void insertIntoPriorityQueue(Graph* graph, int distance, int node) {
    PriorityQueue* newElement = createPriorityQueueElement(distance, node);
    if (graph->priorityQueue == NULL) {
        graph->priorityQueue = newElement;
    } else {
        newElement->next = graph->priorityQueue;
        graph->priorityQueue = newElement;
    }
}

// Function to extract the minimum element from the priority queue
PriorityQueue* extractMinFromPriorityQueue(Graph* graph) {
    if (graph->priorityQueue == NULL) {
        return NULL;
    }
    PriorityQueue* minElement = graph->priorityQueue;
    graph->priorityQueue = graph->priorityQueue->next;
    minElement->next = NULL;
    return minElement;
}

// Function to create a graph
Graph* createGraph(int numNodes) {
    Graph* graph = (Graph*) malloc(sizeof(Graph));
    graph->numNodes = numNodes;
    graph->adjLists = (Edge**) calloc(numNodes, sizeof(Edge*));
    graph->priorityQueue = NULL;
    graph->distances = (int*) calloc(numNodes, sizeof(int));
    graph->visited = (int*) calloc(numNodes, sizeof(int));
    
    for (int i = 0; i < numNodes; i++) {
        graph->distances[i] = INF;
        graph->visited[i] = 0;
    }
    
    return graph;
}

// Function to add an edge to the graph
void addEdge(Graph* graph, int src, int dest, int weight) {
    Edge* newEdge = createEdge(dest, weight);
    newEdge->next = graph->adjLists[src];
    graph->adjLists[src] = newEdge;
}

// Function to implement Dijkstra's algorithm
void dijkstra(Graph* graph, int src) {
    graph->distances[src] = 0;
    insertIntoPriorityQueue(graph, 0, src);
    
    while (graph->priorityQueue != NULL) {
        PriorityQueue* currentElement = extractMinFromPriorityQueue(graph);
        int currentNode = currentElement->node;
        int currentDistance = currentElement->distance;
        
        if (graph->visited[currentNode]) {
            free(currentElement);
            continue;
        }
        
        graph->visited[currentNode] = 1;
        
        Edge* temp = graph->adjLists[currentNode];
        while (temp != NULL) {
            int neighbor = temp->dest;
            int weight = temp->weight;
            
            if (!graph->visited[neighbor] && graph->distances[neighbor] > graph->distances[currentNode] + weight) {
                graph->distances[neighbor] = graph->distances[currentNode] + weight;
                insertIntoPriorityQueue(graph, graph->distances[neighbor], neighbor);
            }
            
            temp = temp->next;
        }
        
        free(currentElement);
    }
}

// Example usage
int main() {
    FILE *file = fopen("../../graph_input.txt", "r");
    if (!file) {
        return 1;
    }

    int numNodes = 1000; // Supondo que o número máximo de nós seja 1000, ajuste se necessário
    Graph* graph = createGraph(numNodes);

    int src, dest, weight;
    while (fscanf(file, "%d %d %d", &src, &dest, &weight) == 3) {
        if (src >= 0 && src < numNodes && dest >= 0 && dest < numNodes) {
            addEdge(graph, src, dest, weight);
            // Se o grafo for não dirigido, também adiciona a aresta inversa:
            // addEdge(graph, dest, src, weight);
        }
    }

    fclose(file);

    dijkstra(graph, 0);

    return 0;
}

File: Python/dijkstra/deepseek-r1-distill-llama-70b.py
graph = {
    'A': [('B', 1), ('C', 4)],
    'B': [('A', 1), ('C', 2), ('D', 5)],
    'C': [('A', 4), ('B', 2), ('D', 1)],
    'D': [('B', 5), ('C', 1)]
}

import heapq

def dijkstra(graph, start_node):
    # Initialize distances to all nodes as infinity except the start node
    distances = {node: float('infinity') for node in graph}
    distances[start_node] = 0
    
    # Priority queue initialized with (distance, node)
    priority_queue = [(0, start_node)]
    
    # Set to keep track of visited nodes
    visited = set()
    
    while priority_queue:
        current_distance, current_node = heapq.heappop(priority_queue)
        
        # Skip if the node has already been processed
        if current_node in visited:
            continue
        
        visited.add(current_node)
        
        # Explore neighbors
        for neighbor, weight in graph[current_node]:
            distance = current_distance + weight
            
            # Only update if a shorter path is found
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(priority_queue, (distance, neighbor))
    
    return distances

# Example usage:
graph = {
    'A': [('B', 1), ('C', 4)],
    'B': [('A', 1), ('C', 2), ('D', 5)],
    'C': [('A', 4), ('B', 2), ('D', 1)],
    'D': [('B', 5), ('C', 1)]
}

def main():
    import sys
    input_path = "graph_input.txt" if len(sys.argv) < 2 else sys.argv[1]

    try:
        with open(input_path, 'r') as f:
            edges = [line.strip().split() for line in f if line.strip()]
    except FileNotFoundError:
        return

    graph = {}
    for u, v, w in edges:
        u, v, w = int(u), int(v), int(w)
        if u not in graph:
            graph[u] = []
        if v not in graph:
            graph[v] = []
        graph[u].append((v, w))
        graph[v].append((u, w))  

    start_node = int(edges[0][0]) 

    distances = dijkstra(graph, start_node)

if __name__ == "__main__":
    main()
